### PR TITLE
Improve (hopefully) how HybridGibbs uses samplers state and history

### LINKED ISSUE #483: Acc rate not stored in HybridGibbs sampler
**Description**

Currently the acceptance rate is not maintained in the HybridGibbs sampler. This means that tuning is not happening using the full acc rate of the chain.

## DoD:
- [x] HybridGibbs ensures samplers store acc rate while sampling
- [x] HybridGibbs maintains state and history of samplers
- [x] unit tests are implemented to check the above mentioned behavior

### PR DESCRIPTION
**Pull request by nabriis**
Closes #483.

This PR makes it such that HybridGibbs will extract each of its samplers state and history, then reinitialize the sampler with a new target, then set the sampler state and history from earlier.

This reason we have to do this is because changing .target will not always correctly set all internal variables of the samplers that depend on the target. 

Also thanks to Silja, ensured that acc rate is stored correctly in HybridGibbs.

## Suggestion for review if method works
Use CWMH and compare with HybridGibbs using MH. Are the results similar? Test both warmup and sampling.

**Silja reviewed and tested. Result OK.**

### CODE REVIEW

**Comment by siljachr** [comment]
The MH sampler did not append to self._acc, and therefore self._acc stayed at the initial value 1, which meant that the tune method increased the step length to the maximum (i.e. 1). I have pushed a small change that fixes this. 

In line 220 of cuqi/experimental/mcmc/_gibbs, the skip_len of the tune method is hardcoded to 1, meaning that the tuning of the scale parameter is always based on only the latest self._acc, i.e. either 0 or 1. This is not sensible for tuning. I suggest either using skip_len = max(int(0.1 * Nb), 1) which is the default in the warmup method in cuqi/experimental/mcmc/_sampler. Or letting it be user defined, again similar to the implementation in the warmup method in cuqi/experimental/mcmc/_sampler.

**Comment by siljachr** [comment]
> The MH sampler did not append to self._acc, and therefore self._acc stayed at the initial value 1, which meant that the tune method increased the step length to the maximum (i.e. 1). I have pushed a small change that fixes this.
> 
> In line 220 of cuqi/experimental/mcmc/_gibbs, the skip_len of the tune method is hardcoded to 1, meaning that the tuning of the scale parameter is always based on only the latest self._acc, i.e. either 0 or 1. This is not sensible for tuning. I suggest either using skip_len = max(int(0.1 * Nb), 1) which is the default in the warmup method in cuqi/experimental/mcmc/_sampler. Or letting it be user defined, again similar to the implementation in the warmup method in cuqi/experimental/mcmc/_sampler.

I fixed an issue concerning how I first implemented appendence to self._acc. Now implemented in HybridGibbs instead of MH so it is also compatible with regular sampling (non-Gibbs), which already has  appendence to self._acc implemented. 

I have added the option of setting the tune_freq for HybridGibbs sampling, so it is no longer hard-coded. The implementation is equivalent to tune_freq in cuqi/experimental/mcmc/_sampler. 

I have no further comments or concerns related to storing and using the acceptance rate in HybridGibbs :)

File: tests/zexperimental/test_mcmc.py [hunk 1]
```diff
@@ -823,3 +823,51 @@ def test_HybridGibbs_initial_point_setting():
...
+    sampler.warmup(Nb).sample(Ns)
+
+    # Check that the state is different after running the sampler again
+    for key in sampler.samplers.keys():
+        new_state = sampler.samplers[key].get_state()
+        # If sampler has no accepeted samples in last run (rare), the state should not be updated
+        if (np.sum(sampler.samplers[key].get_history()["history"]["_acc"][Ns+Nb+1:]) == 0):
+            assert sampler_states[key] == new_state, f"Sampler {key} state is updated after Gibbs sampling, but no accepted samples. State: \n {new_state}"
+        else:
+            assert sampler_states[key] != new_state, f"Sampler {key} state is not updated after Gibbs sampling. State: \n {new_state}"
```
Comment (amal-ghamdi) [inline tests/zexperimental/test_mcmc.py:872]: I am a bit confused about the relation between the assertion condition and the message. For the first assertion, the state seems to be not updated (only the history), hence `sampler_states[key] == new_state`
For the second case. The state is updated, hence `sampler_states[key] != new_state`? does that make sense? The messages seem to convey the opposite.

**Review by amal-ghamdi — CHANGES_REQUESTED** [review]
Super!, many thanks @nabriis and @siljachr. Looks great.

I have a few comments. Feel free to address as you see fit.

**Review by chaozg — COMMENTED** [review]
Thanks @nabriis and @siljachr for this nice PR. I just have a tiny reminder that the example in the docstring of `HybridGibbs` nees a bit update as it still follows the old way of sampling. I mean `samples = sampler.sample(Ns=1000, Nb=200)` [here](https://github.com/CUQI-DTU/CUQIpy/blob/1322f5fd36715394d8e8d945644fb2a618218c80/cuqi/experimental/mcmc/_gibbs.py#L93).

File: tests/zexperimental/test_mcmc.py (hunk 1 above)
Comment (nabriis) [reply tests/zexperimental/test_mcmc.py]: I updated the description and refactored the test. Hopefully it is now more clear.

File: cuqi/experimental/mcmc/_gibbs.py [hunk 2]
```diff
@@ -182,33 +215,58 @@ def step(self):
...
-            sampler.initial_point = self.current_samples[par_name]
-            if hasattr(sampler, 'initial_scale'): sampler.initial_scale = sampler.scale
+            # Instead of simply changing the target of the sampler, we reinitialize it.
+            # This is to ensure that all internal variables are set to match the new target.
+            # To return the sampler to the old state and history, we first extract the state and history
+            # before reinitializing the sampler and then set the state and history back to the sampler
+
+            # Extract state and history from sampler
+            if isinstance(sampler, NUTS): # Special case for NUTS as it is not playing nice with get_state and get_history
+                sampler.initial_point = sampler.current_point
```
[the remark this answers is not in the document]
Comment (nabriis) [reply cuqi/experimental/mcmc/_gibbs.py:225]: Perhaps yes. I suggest we create an issue for it though. This does not dimish existing behavior of the NUTS sampler.

**Review by chaozg — APPROVED** [review]
LGTM. Regarding the issue of the example at the docstring of `HybridGibbs`, which is of course out the scope of this PR, I took the liberty of creating an issue to track it: #511

**Review by amal-ghamdi — CHANGES_REQUESTED** [review]
Thank you @nabriis, I just added one comment regarding printing a warning for NUTS.

File: cuqi/experimental/mcmc/_gibbs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nabriis) [reply cuqi/experimental/mcmc/_gibbs.py:225]: Created #512  and added warning

**Review by amal-ghamdi — APPROVED** [review]
LGTM!

**Pull request merged**